packages/muelu/src/Utils/MueLu_UtilitiesBase_decl.hpp

```diff
@@ -129,6 +135,95 @@ namespace MueLu {
       return rcp(new CrsMatrixWrap(Op));
     }
 
+    /*! @brief Threshold a matrix
+
+    Returns matrix filtered with a threshold value.
+
+    NOTE -- it's assumed that A has been fillComplete'd.
+    */
+    static RCP<CrsMatrixWrap> GetThresholdedMatrix(const RCP<Matrix>& Ain, const Scalar threshold, const bool keepDiagonal=true, const GlobalOrdinal expectedNNZperRow=-1) {
+
+      RCP<const Map> rowmap = Ain->getRowMap();
+      RCP<const Map> colmap = Ain->getColMap();
+      RCP<CrsMatrixWrap> Aout = rcp(new CrsMatrixWrap(rowmap, expectedNNZperRow <= 0 ? Ain->getGlobalMaxNumRowEntries() : expectedNNZperRow));
+      // loop over local rows
+      for(size_t row=0; row<Ain->getLocalNumRows(); row++)
+      {
+        size_t nnz = Ain->getNumEntriesInLocalRow(row);
+
+        Teuchos::ArrayView<const LocalOrdinal> indices;
+        Teuchos::ArrayView<const Scalar> vals;
+        Ain->getLocalRowView(row, indices, vals);
+
+        TEUCHOS_TEST_FOR_EXCEPTION(Teuchos::as<size_t>(indices.size()) != nnz, Exceptions::RuntimeError, "MueLu::ThresholdAFilterFactory::Build: number of nonzeros not equal to number of indices? Error.");
+
+        Teuchos::ArrayRCP<GlobalOrdinal> indout(indices.size(),Teuchos::ScalarTraits<GlobalOrdinal>::zero());
+        Teuchos::ArrayRCP<Scalar> valout(indices.size(),Teuchos::ScalarTraits<Scalar>::zero());
+        size_t nNonzeros = 0;
+        if (keepDiagonal) {
+          GlobalOrdinal glbRow = rowmap->getGlobalElement(row);
+          LocalOrdinal lclColIdx = colmap->getLocalElement(glbRow);
+          for(size_t i=0; i<(size_t)indices.size(); i++) {
+            if(Teuchos::ScalarTraits<Scalar>::magnitude(vals[i]) > Teuchos::ScalarTraits<Scalar>::magnitude(threshold) || indices[i]==lclColIdx) {
+              indout[nNonzeros] = colmap->getGlobalElement(indices[i]); // LID -> GID (column)
```

Comment:
(Not at the core of this PR, but maybe easy to do.) What about using [`Xpetra::Map::getLocalElementList()`](https://docs.trilinos.org/dev/packages/xpetra/doc/html/classXpetra_1_1Map.html#af8dd381e64b40f3015ca52795cae78f3) to get the list only once and then grab just local values to ask for GIDs?